tests/unit/test_protocols.py

```diff
@@ -279,32 +292,55 @@ def _fixup_parsed_result(parsed):
         for key in error_keys:
             if key not in ['Code', 'Message']:
                 del parsed['Error'][key]
+    # 5. Special float types. In the protocol test suite, certain special float
+    # types are represented as strings: "Infinity", "-Infinity", and "NaN".
+    # However, we parse these values as actual floats types, so we need to convert
+    # them back to their string representation.
+    parsed = _convert_special_floats_to_string(parsed)
     return parsed
 
 
-def _convert_bytes_to_str(parsed):
-    if isinstance(parsed, dict):
-        new_dict = {}
-        for key, value in parsed.items():
-            new_dict[key] = _convert_bytes_to_str(value)
-        return new_dict
-    elif isinstance(parsed, bytes):
-        return parsed.decode('utf-8')
-    elif isinstance(parsed, list):
-        new_list = []
-        for item in parsed:
-            new_list.append(_convert_bytes_to_str(item))
-        return new_list
+def _convert(obj, conversion_funcs):
+    if isinstance(obj, dict):
+        return {k: _convert(v, conversion_funcs) for k, v in obj.items()}
+    elif isinstance(obj, list):
+        return [_convert(item, conversion_funcs) for item in obj]
     else:
-        return parsed
+        for conv_type, conv_func in conversion_funcs:
+            if isinstance(obj, conv_type):
+                return conv_func(obj)
+    return obj
+
+
+def _bytes_to_str(value):
+    if isinstance(value, bytes):
+        return value.decode('utf-8')
+    return value
+
+
+def _convert_bytes_to_str(parsed):
+    return _convert(parsed, [(bytes, _bytes_to_str)])
+
+
+def _special_floats_to_str(value):
+    if isinstance(value, float):
+        if value in [float('Infinity'), float('-Infinity')] or math.isnan(
+            value
+        ):
```

Comment:
Is there something unique about `NaN` that we don't use `float('NaN')` for the check like we do inf?